File: android/src/main/java/expo/modules/breathing/MorphingBlobView.kt
package expo.modules.breathing

import androidx.compose.foundation.Canvas
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.DrawScope
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.tan

@Composable
fun MorphingBlobView(
    baseRadius: Float,
    pointCount: Int,
    offsets: List<Double>,
    colors: List<Color>,
    innerColor: Color,
    showInnerBlob: Boolean,
    modifier: Modifier = Modifier
) {
    Canvas(modifier = modifier) {
        val size = min(this.size.width, this.size.height)
        val center = Offset(this.size.width / 2f, this.size.height / 2f)
        val effectiveRadius = baseRadius * size / 2f

        // Draw main blob
        val mainPath = createBlobPath(center, effectiveRadius, pointCount, offsets)
        drawPath(
            path = mainPath,
            brush = Brush.radialGradient(
                colors = colors,
                center = center,
                radius = effectiveRadius
            )
        )

        // Draw inner blob for depth
        if (showInnerBlob) {
            val innerPath = createBlobPath(
                center = center,
                baseRadius = effectiveRadius * 0.6f,
                pointCount = pointCount,
                offsets = offsets.map { it * 0.7 }
            )
            drawPath(
                path = innerPath,
                color = innerColor
            )
        }
    }
}

private fun DrawScope.createBlobPath(
    center: Offset,
    baseRadius: Float,
    pointCount: Int,
    offsets: List<Double>
): Path {
    val path = Path()
    if (pointCount < 3) return path

    // Generate points around the circle with offsets
    val points = mutableListOf<Offset>()
    for (i in 0 until pointCount) {
        val angle = (i.toDouble() / pointCount) * 2 * PI - PI / 2
        val offset = if (i < offsets.size) offsets[i] else 0.0
        val radius = baseRadius * (1.0 + offset).toFloat()

        points.add(
            Offset(
                x = center.x + (cos(angle) * radius).toFloat(),
                y = center.y + (sin(angle) * radius).toFloat()
            )
        )
    }

    // Tangent coefficient for smooth cubic bezier curves
    val tangentCoeff = ((4.0 / 3.0) * tan(PI / (2.0 * pointCount))).toFloat()

    // Start at first point
    path.moveTo(points[0].x, points[0].y)

    // Draw cubic bezier curves between each pair of points
    for (i in 0 until pointCount) {
        val current = points[i]
        val next = points[(i + 1) % pointCount]

        // Calculate angles for tangent directions
        val currentAngle = (i.toDouble() / pointCount) * 2 * PI - PI / 2
        val nextAngle = (((i + 1) % pointCount).toDouble() / pointCount) * 2 * PI - PI / 2

        // Current point's radius and tangent
        val currentOffset = if (i < offsets.size) offsets[i] else 0.0
        val currentRadius = baseRadius * (1.0 + currentOffset).toFloat()
        val currentTangentLength = currentRadius * tangentCoeff

        // Next point's radius and tangent
        val nextOffset = if ((i + 1) % pointCount < offsets.size) offsets[(i + 1) % pointCount] else 0.0
        val nextRadius = baseRadius * (1.0 + nextOffset).toFloat()
        val nextTangentLength = nextRadius * tangentCoeff

        // Control points are perpendicular to radial direction
        val control1 = Offset(
            x = current.x + (cos(currentAngle + PI / 2) * currentTangentLength).toFloat(),
            y = current.y + (sin(currentAngle + PI / 2) * currentTangentLength).toFloat()
        )

        val control2 = Offset(
            x = next.x + (cos(nextAngle - PI / 2) * nextTangentLength).toFloat(),
            y = next.y + (sin(nextAngle - PI / 2) * nextTangentLength).toFloat()
        )

        path.cubicTo(
            control1.x, control1.y,
            control2.x, control2.y,
            next.x, next.y
        )
    }

    path.close()
    return path
}

File: ios/Breathing/MorphingBlobView.swift
import SwiftUI

struct MorphingBlobView: View {
    let baseRadius: CGFloat
    let pointCount: Int
    let offsets: [Double]
    let colors: [Color]
    let innerColor: Color
    let showInnerBlob: Bool
    var elapsedTime: Double = 0

    var body: some View {
        GeometryReader { geometry in
            let size = min(geometry.size.width, geometry.size.height)
            let center = CGPoint(x: size / 2, y: size / 2)
            let effectiveRadius = baseRadius * size / 2

            ZStack {
                // Main blob with linear gradient fill (like OrbView)
                MorphingBlobShape(
                    center: center,
                    baseRadius: effectiveRadius,
                    pointCount: pointCount,
                    offsets: offsets
                )
                .fill(
                    LinearGradient(
                        colors: colors,
                        startPoint: .bottom,
                        endPoint: .top
                    )
                )

                // Inner glow overlay (like OrbView's realisticInnerGlows)
                MorphingBlobShape(
                    center: center,
                    baseRadius: effectiveRadius,
                    pointCount: pointCount,
                    offsets: offsets
                )
                .stroke(
                    LinearGradient(
                        colors: [.white.opacity(0.6), .clear],
                        startPoint: .bottom,
                        endPoint: .top
                    ),
                    lineWidth: 3
                )
                .blur(radius: 4)
                .blendMode(.plusLighter)

                // Inner animated blob like OrbView's WavyBlobView
                if showInnerBlob {
                    InnerWavyBlob(
                        center: center,
                        baseRadius: effectiveRadius * 0.55,
                        color: innerColor,
                        elapsedTime: elapsedTime
                    )
                    .blur(radius: 10)
                    .blendMode(.plusLighter)
                }
            }
        }
    }
}

// Inner blob with time-based animation like WavyBlobView
struct InnerWavyBlob: View {
    let center: CGPoint
    let baseRadius: CGFloat
    let color: Color
    let elapsedTime: Double

    private let pointCount = 6
    private let loopDuration: Double = 4.0

    var body: some View {
        Canvas { context, size in
            let angle = (elapsedTime.truncatingRemainder(dividingBy: loopDuration) / loopDuration) * 2 * .pi

            var path = Path()

            // Generate points with time-based smooth movement
            var points: [CGPoint] = []
            for i in 0..<pointCount {
                let baseAngle = (Double(i) / Double(pointCount)) * 2 * .pi
                let phaseOffset = Double(i) * .pi / 3

                // Smooth oscillating offsets
                let radiusOffset = sin(angle * 2 + phaseOffset) * 0.15 + cos(angle * 1.5 + phaseOffset * 0.7) * 0.1
                let angleOffset = cos(angle + phaseOffset) * 0.1

                let r = baseRadius * (1.0 + radiusOffset)
                let a = baseAngle + angleOffset

                points.append(CGPoint(
                    x: center.x + cos(a) * r,
                    y: center.y + sin(a) * r
                ))
            }

            // Start path
            path.move(to: points[0])

            // Create smooth curves between points
            let handleLength = baseRadius * 0.5

            for i in 0..<pointCount {
                let current = points[i]
                let next = points[(i + 1) % pointCount]

                let currentAngle = atan2(current.y - center.y, current.x - center.x)
                let nextAngle = atan2(next.y - center.y, next.x - center.x)

                let control1 = CGPoint(
                    x: current.x + cos(currentAngle + .pi / 2) * handleLength,
                    y: current.y + sin(currentAngle + .pi / 2) * handleLength
                )

                let control2 = CGPoint(
                    x: next.x + cos(nextAngle - .pi / 2) * handleLength,
                    y: next.y + sin(nextAngle - .pi / 2) * handleLength
                )

                path.addCurve(to: next, control1: control1, control2: control2)
            }

            path.closeSubpath()
            context.fill(path, with: .color(color))
        }
    }
}

struct MorphingBlobShape: Shape {
    var center: CGPoint
    var baseRadius: CGFloat
    var pointCount: Int
    var offsets: [Double]

    var animatableData: AnimatableVector {
        get { AnimatableVector(values: offsets) }
        set { offsets = newValue.values }
    }

    func path(in rect: CGRect) -> Path {
        var path = Path()

        guard pointCount >= 3 else { return path }

        // Generate points around the circle with offsets
        var points: [CGPoint] = []
        for i in 0..<pointCount {
            let angle = (Double(i) / Double(pointCount)) * 2 * .pi - .pi / 2
            let offset = i < offsets.count ? offsets[i] : 0
            let radius = baseRadius * (1.0 + offset)

            points.append(CGPoint(
                x: center.x + CGFloat(cos(angle)) * radius,
                y: center.y + CGFloat(sin(angle)) * radius
            ))
        }

        // Tangent coefficient for smooth cubic bezier curves
        let tangentCoeff = CGFloat((4.0 / 3.0) * tan(.pi / (2.0 * Double(pointCount))))

        // Start at first point
        path.move(to: points[0])

        // Draw cubic bezier curves between each pair of points
        for i in 0..<pointCount {
            let current = points[i]
            let next = points[(i + 1) % pointCount]

            let currentAngle = (Double(i) / Double(pointCount)) * 2 * .pi - .pi / 2
            let nextAngle = (Double((i + 1) % pointCount) / Double(pointCount)) * 2 * .pi - .pi / 2

            let currentOffset = i < offsets.count ? offsets[i] : 0
            let currentRadius = baseRadius * CGFloat(1.0 + currentOffset)
            let currentTangentLength = currentRadius * tangentCoeff

            let nextOffset = (i + 1) % pointCount < offsets.count ? offsets[(i + 1) % pointCount] : 0
            let nextRadius = baseRadius * CGFloat(1.0 + nextOffset)
            let nextTangentLength = nextRadius * tangentCoeff

            let control1 = CGPoint(
                x: current.x + CGFloat(cos(currentAngle + .pi / 2)) * currentTangentLength,
                y: current.y + CGFloat(sin(currentAngle + .pi / 2)) * currentTangentLength
            )

            let control2 = CGPoint(
                x: next.x + CGFloat(cos(nextAngle - .pi / 2)) * nextTangentLength,
                y: next.y + CGFloat(sin(nextAngle - .pi / 2)) * nextTangentLength
            )

            path.addCurve(to: next, control1: control1, control2: control2)
        }

        path.closeSubpath()
        return path
    }
}

struct AnimatableVector: VectorArithmetic {
    var values: [Double]

    static var zero: AnimatableVector {
        AnimatableVector(values: [])
    }

    static func + (lhs: AnimatableVector, rhs: AnimatableVector) -> AnimatableVector {
        let maxCount = max(lhs.values.count, rhs.values.count)
        var result = [Double](repeating: 0, count: maxCount)
        for i in 0..<maxCount {
            let l = i < lhs.values.count ? lhs.values[i] : 0
            let r = i < rhs.values.count ? rhs.values[i] : 0
            result[i] = l + r
        }
        return AnimatableVector(values: result)
    }

    static func - (lhs: AnimatableVector, rhs: AnimatableVector) -> AnimatableVector {
        let maxCount = max(lhs.values.count, rhs.values.count)
        var result = [Double](repeating: 0, count: maxCount)
        for i in 0..<maxCount {
            let l = i < lhs.values.count ? lhs.values[i] : 0
            let r = i < rhs.values.count ? rhs.values[i] : 0
            result[i] = l - r
        }
        return AnimatableVector(values: result)
    }

    mutating func scale(by rhs: Double) {
        for i in 0..<values.count {
            values[i] *= rhs
        }
    }

    var magnitudeSquared: Double {
        values.reduce(0) { $0 + $1 * $1 }
    }
}
